### PR TITLE
Add favicons to external link previews

### PR DESCRIPTION
**Pull request by ibrkhalil**
This PR scrapes the HTML of the link preview searching for the favicon, If found it returns it's dataUri to mobile for the message input level link preview, When sending the message to Waku we store the favicon data as a media server URL.

Required for https://github.com/status-im/status-mobile/pull/18484#pullrequestreview-1890732018

Please note that fetching all favicons is not supported at the moment, Will create follow-ups on status-mobile to improve this logic in follow-ups in the near future to support more.

### CODE REVIEW

File: protocol/common/message_linkpreview.go
Comment (igor-sirotin) [inline protocol/common/message_linkpreview.go]: Why not `favicon` in json? 😄 
```suggestion
	Favicon     string                         `json:"favicon,omitempty"`
```

**Review by igor-sirotin — CHANGES_REQUESTED** [review]
So `Favicon` is just a URL pointing to some website, did I get it right?

We want to fetch the Favicon and attach it to the message. Please, take a look how `Thumbnail` is implemented.

File: images/decode.go
Comment (ibrkhalil) [inline images/decode.go]: According to Identification section [here](http://fileformats.archiveteam.org/wiki/ICO)

File: protocol/common/message_linkpreview.go
Comment (ibrkhalil) [reply protocol/common/message_linkpreview.go]: Done, Can you kindly take another look?

File: images/decode.go
Comment (igor-sirotin) [inline images/decode.go]: Why do we need such an alias? Also, `Is()` is not quite a self-explaining name 🤔

Comment (igor-sirotin) [reply images/decode.go]: So there's 3 functions doing the same thing: `Is`, `IsSVG` and `IsSvg`. Why? 🤔

**Review by igor-sirotin — CHANGES_REQUESTED** [review]
I think you're missing a part with media server 🤔

File: images/decode.go
Comment (ibrkhalil) [reply images/decode.go]: I copied the core code of a package, But I didn't refactor it.
Fixing.

File: images/test_data.go
Comment (ibrkhalil) [inline images/test_data.go:9]: Had to increase the size of this dummy data array with more dummy data, To make GetSVG not throw indexOutOfBound error.

File: appdatabase/migrationsprevnodecfg/bindata.go
Comment (igor-sirotin) [inline appdatabase/migrationsprevnodecfg/bindata.go]: This shouldn't be here. Please rebase and re-run `make generate` with `nix-shell`: 
https://www.notion.so/status-go-make-generate-d700f948e9a446f8a3a3b859e8b00a5f?pvs=4

File: protocol/common/message_linkpreview.go
Comment (igor-sirotin) [inline protocol/common/message_linkpreview.go]: hmmm. I suppose it should be of `LinkPreviewThumbnail` type. 🤔

File: protocol/linkpreview_unfurler_opengraph.go
Comment (igor-sirotin) [inline protocol/linkpreview_unfurler_opengraph.go]:
```suggestion
			if t.Data != "link" {
			  continue
			}
		    LinkContainsFavicon := false
		    for _, attribute := range t.Attr {
			    if attribute.Val == "icon" || attribute.Val == "shortcut icon" {
				    LinkContainsFavicon = true
			    }
		    }
		    for _, attribute := range t.Attr {
			    if LinkContainsFavicon && (strings.Contains(attribute.Val, ".ico") || strings.Contains(attribute.Val, ".png") || strings.Contains(attribute.Val, ".svg")) {
				    return attribute.Val
			    }
		    }
```

**Review by igor-sirotin — CHANGES_REQUESTED** [review]
LGTM, main concern is the migration files 🙂

**Review by igor-sirotin — APPROVED** [review]
Thank you!

**Comment by igor-sirotin** [comment]
I suggest we wait for a review from @ilmotta as well 🙂

File: images/decode.go
Comment (ilmotta) [inline images/decode.go]: Typo `Is` -> `IsSVG`

File: images/decode.go
Comment (ilmotta) [inline images/decode.go:29]: Missing unit test.

File: images/decode.go
Comment (ilmotta) [inline images/decode.go:184]: Missing unit test.

File: protocol/common/message_linkpreview.go
Comment (ilmotta) [inline protocol/common/message_linkpreview.go]: We could be more specific in the error message since now the data URI could come from fetching a favicon or a thumbnail. E.g. `"could not get thumbnail data URI payload"`

File: protocol/messenger_linkpreview_test.go
Comment (ilmotta) [inline protocol/messenger_linkpreview_test.go:201]: You could cover more cases in the test instead of just the simplest happy path.

Example:

- Any combination of `ico`, `svg`, and `png` links
- Link tags without `icon` or `shortcut icon`
- Other tags mixed inside `head`.

File: server/handlers_linkpreview.go
Comment (ilmotta) [inline server/handlers_linkpreview.go]: I don't think we should duplicate this much. This function is 99% identical to `getThumbnailPayload`, except for the part `payload = p.FaviconPayload`.

File: server/handlers_linkpreview.go
Comment (ilmotta) [inline server/handlers_linkpreview.go]: Same as my comment above. We shouldn't copy & paste this much. The only difference with `handleLinkPreviewThumbnail` is that you need to call `getFaviconPayload` instead of `getThumbnailPayload`. You can create a common function that receives a function fulfilling the signature of both (which are identical).

File: images/decode.go
Comment (ilmotta) [inline images/decode.go:40]: Once you cover this function with a unit test, it would be a good idea to use non-trivial SVG(s) to see if the complicated regex `svgRegex` is actually good enough. I don't know if any LLM recommended that to you, but I had chatgpt sharing blatantly wrong regexes to me up to a point I gave up and wrote by hand.

File: protocol/linkpreview_unfurler_opengraph.go
Comment (ilmotta) [inline protocol/linkpreview_unfurler_opengraph.go]: You can break out of this loop once you found there's a favicon link.

File: protocol/linkpreview_unfurler_opengraph.go
Comment (ilmotta) [inline protocol/linkpreview_unfurler_opengraph.go]: I don't understand why we need to loop twice. Another problem is that the first loop could have found that there is in fact a favicon link, but it doesn't mean it's the same as the one currently under iteration in the second loop. It seems to me to be more precise to check here `attribute.Val` is either `icon` or `shortcut icon`, and then we can also eliminate the first loop.

File: protocol/linkpreview_unfurler_opengraph.go
Comment (ilmotta) [inline protocol/linkpreview_unfurler_opengraph.go]: Nitpick: the function name `fetchThumbnail` could be renamed. Something like `fetchImage` would be more appropriate because `thumbnail` isn't a word used as a synonym for `favicon`.

File: protocol/linkpreview_unfurler_opengraph.go
Comment (ilmotta) [inline protocol/linkpreview_unfurler_opengraph.go:139]: What if the image is an `ico`? Would you get the dimensions back? If so, we're not storing them in the protocol (only `favicon_payload`), so is there any need to get `ico`'s dimensions?

File: images/decode.go
Comment (ibrkhalil) [reply images/decode.go:40]: I got it from a lib that did svg type check on bytes.

Comment (ibrkhalil) [reply images/decode.go:40]: Also to be honest, I don't really know how non-trivial SVG(s) can be
Since it's markup it's so hard to have multiple ways to do SVGs.
If you have any examples that'd help me a lot.

File: protocol/linkpreview_unfurler_opengraph.go
Comment (ibrkhalil) [reply protocol/linkpreview_unfurler_opengraph.go]: When we encounter a start tag, We loop the first loop to check all attributes of that start tag and make sure it's actually a favicon

The second loop we double check if the favicon resources is actually one of our supported types and then we return it.

Comment (ibrkhalil) [reply protocol/linkpreview_unfurler_opengraph.go]: Notice the `case html.StartTagToken:`
The tag is the same in both loops.
But we don't know for a fact in what order the HTML attributes are added to the start tag, That's why I loop through all of them first to make sure it's in fact a link tag that's meant to be a favicon.

File: protocol/linkpreview_unfurler_opengraph.go
Comment (ibrkhalil) [reply protocol/linkpreview_unfurler_opengraph.go]: Done

File: protocol/common/message_linkpreview.go
Comment (ibrkhalil) [reply protocol/common/message_linkpreview.go]: Done

File: images/decode.go
Comment (ibrkhalil) [reply images/decode.go:184]: Done

File: images/decode.go
Comment (ibrkhalil) [reply images/decode.go:29]: Done

File: images/decode.go
Comment (ibrkhalil) [reply images/decode.go]: Done

File: protocol/linkpreview_unfurler_opengraph.go
Comment (ilmotta) [reply protocol/linkpreview_unfurler_opengraph.go]: Thanks for the explanation @ibrkhalil.

I think we can improve this implementation. We don't need to loop twice and the detection of the icon could improve.

The following code passes your test and another one I experimented with a much more complicated `head` tag.

```go
func GetFavicon(bodyBytes []byte) string {
	htmlTokens := html.NewTokenizer(bytes.NewBuffer(bodyBytes))
loop:
	for {
		tt := htmlTokens.Next()
		switch tt {
		case html.ErrorToken:
			break loop
		case html.StartTagToken:
			t := htmlTokens.Token()
			if t.Data != "link" {
				continue
			}

			isIcon := false
			href := ""
			for _, attr := range t.Attr {
				k := attr.Key
				v := attr.Val
				if k == "rel" && (v == "icon" || v == "shortcut icon") {
					isIcon = true
				} else if k == "href" &&
					(strings.Contains(v, ".ico") ||
						strings.Contains(v, ".png") ||
						strings.Contains(v, ".svg")) {
					href = v
				}
			}

			if isIcon && href != "" {
				return href
			}
		}
	}
	return ""
}
```

The example above is more precise because it only tries to detect an icon if the attribute key is `rel`. The other thing I noticed is that some websites report the `type` of the link. For example, GitHub uses `<link rel="icon" type="image/svg+xml" href="https://...">`. Not all websites do this though, but we could check first based on the type tag if it's available. Using `contains` is not very precise, but it gets the job done I suppose.

File: images/decode.go
Comment (ilmotta) [reply images/decode.go:40]: I would use any reasonably complex SVG or two you could find on the internet and commit those in the test (be careful with copyright), but also double check a few other SVGs just to see if the test is correct (but I wouldn't commit in the code, just check and delete).

You could also get the SVG favicons from a top 5+ list of known websites to double-check as well (if you haven't already).

File: protocol/linkpreview_unfurler_opengraph.go
Comment (ibrkhalil) [reply protocol/linkpreview_unfurler_opengraph.go:139]: The thing is, attempting to fetch SVGs dimensions breaks the entire thing
Because SVGs are markup and they need to be rendered in order to fetch their dimension
But you're right, I should omit calculating dimensions for favicons completely.

File: protocol/linkpreview_unfurler_opengraph.go
Comment (ibrkhalil) [reply protocol/linkpreview_unfurler_opengraph.go]: I copied this implementation, Thanks Icaro!

**Review by ilmotta — APPROVED** [review]
PR LGTM @ibrkhalil. Thank you for working in all the review feedback 💯

**Pull request merged**
